extensions/amp-viz-vega/0.1/amp-viz-vega.js

```diff
@@ -0,0 +1,166 @@
+import {toggle} from '../../../src/style';
+
+/** @const */
+const EXPERIMENT = 'amp-viz-vega';
+
+export class AmpVizVega extends AMP.BaseElement {
+
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /** @private {?Element} */
+    this.container_ = null;
+
+    /** @private {?JSONType} */
+    this._spec = null;
+
+    /** @private {?string} */
+    this._specUrl = null;
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return isLayoutSizeDefined(layout);
+  }
+
+  /**
+   * Create the vega container.
+   * Called lazily in the first `#layoutCallback`.
+   */
+  initialize_() {
+    if (this.container_) {
+      return;
+    }
+
+    this.container_ = this.element.ownerDocument.createElement('div');
+    this._specUrl = this.element.getAttribute('spec-url');
+
+    this.applyFillContent(this.container_, true);
+    this.element.appendChild(this.container_);
+  }
+
+  /** @override */
+  buildCallback() {
+    if (!isExperimentOn(this.win, EXPERIMENT)) {
+      dev.warn(EXPERIMENT, `Experiment ${EXPERIMENT} disabled`);
+      toggle(this.element, false);
+      return;
+    }
+  }
+
+  /** @override */
+  layoutCallback() {
+    this.initialize_();
+    return this.loadSpec_().then(() => this.renderGraph_());
+  }
+
+  /**
+   * @return {!Promise}
+   * @private
+   */
+  loadSpec_() {
+    if (this._spec) {
+      return Promise.resolve();
+    }
+    const inlineSpec = this.getInlineSpec();
+    if (!inlineSpec && !this._specUrl) {
+      const err = this.getName_() + ' neither the spec-url attribute nor a' +
+        'valid <script type="application/json"> child was found for Vega spec';
+      return Promise.reject(new Error(err));
+    }
+
+    if (inlineSpec && this._specUrl) {
+      const err = this.getName_() + ' both the spec-url attribute and a valid' +
+        '<script type="application/json"> child were found for Vega spec.' +
+        'Only one way of specifying the spec is allowed.';
+      return Promise.reject(new Error(err));
+    }
+
+    if (this._specUrl) {
+      // TODO(aghassemi): Fetch and validate the spec file.
+      this._spec = {
+        'url': this._specUrl,
+      };
+      return Promise.resolve();
+    }
+
+    if (inlineSpec) {
+      this._spec = inlineSpec;
+      return Promise.resolve();
+    }
+  }
+
+  /**
+   * @return {?string}
+   * @private
+   */
+  getInlineSpec() {
+    let inlineConfig;
+
+    const scripts = childElementsByTag(this.element, 'SCRIPT');
+    if (scripts.length == 1) {
+      const child = scripts[0];
+      if (isJsonScriptTag(child)) {
+        inlineConfig = tryParseJson(scripts[0].textContent, err => {
+          user.error(this.getName_(), 'spec could not be ' +
+            'parsed. Is it in a valid JSON format?', err);
+        });
+      } else {
+        user.error(this.getName_(), 'spec should ' +
+          'be put in a <script type="application/json" tag.');
+      }
+    }
+    return inlineConfig;
```

Comment:
`undefined` is not defined as a possible value by `@return` above. Also, is it worth to show an error if there are two `<script>` tags? Currently this is just ignored.